pyrdp/player/gdi/draw.py

```diff
@@ -0,0 +1,702 @@
+    possible, but GDI specific operations that are not supported by Qt
+    are implemented here.
+
+    Some of the methods are not implemented and will simply ignore the
+    order:
+
+        - NineGrid-related orders
+        - StreamBitmap orders
+        - GDI+ orders
+
+    Note that `CacheBitmapV3` is only used in 32 bits-per-pixel which
+    PyRDP currently does not support due to RDP6.0 compression not
+    being implemented yet.
+
+    If you have a replay which contains any unsupported or untested
+    order, do not hesitate to share it with the project maintainers so
+    that support can be added as required.
+
+    (Make sure that the trace does not contain any sensitive information)
+    """
+
+    def __init__(self, dc: QRemoteDesktop):
+        self.dc = dc
+        self._warned = False
+
+        # Initialize caches.
+        self.bitmaps = BitmapCache()
+        self.brushes = BrushCache()
+        self.palettes = PaletteCache()
+        self.glyphs = GlyphCache()
+
+        self.bounds = None
+
+        screen = dc.screen
+        fallback = QImage(dc.width(), dc.height(), QImage.Format_ARGB32_Premultiplied)
+        fallback.fill(0)
+
+        self.surfaces = {
+            SCREEN_BITMAP_SURFACE: screen,
+            SUBSTITUTE_SURFACE: fallback,
+        }
+        self.activeSurface = SCREEN_BITMAP_SURFACE
+
+    @property
+    def surface(self) -> QImage:
+        """Get the currently active surface."""
+        return self.surfaces[self.activeSurface]
+
+    @property
+    def screen(self) -> QImage:
+        return self.surfaces[SCREEN_BITMAP_SURFACE]
+
+    # Rendering Helpers.
+    def _paint(self, dst: QImage):
+        """Retrieve QPainter for the given surface."""
+        p = QPainter(dst)
+        p.setPen(Qt.NoPen)
+        # Set the bounding rectangle if present.
+        if self.bounds:
+            x = self.bounds.left
+            y = self.bounds.top
+            w = self.bounds.right - x
+            h = self.bounds.bottom - y
+
+            p.setClipRect(x, y, w + 1, h + 1)
+            p.setClipping(True)
+        return p
+
+    def _end(self, p: QPainter):
+        p.end()
+
+    def _brush(self, b: Brush, p: QPainter):
+        """Configure the given brush."""
+        brush = None
+
+        if b.index is not None:  # This is a cached brush.
+            brush = self.brushes.get(b.index)
+        elif b.style == BrushStyle.PATTERN:
+            pm = QPixmap.loadFromData(b.data, _fmt[b.bpp])
+            brush = QBrush(pm)
+        elif b.style == BrushStyle.HATCHED:
+            brush = QBrush(_hs[b.hatch])
+        else:
+            brush = QBrush(_bs[b.style])
+
+        p.setBrush(brush)
+        p.setBrushOrigin(b.x, b.y)
+
+    # Drawing API.
+    def onBounds(self, bounds: Bounds):
+        """Called on bounding rectangle updates."""
+        self.bounds = bounds
+
+    def dstBlt(self, state: DstBlt):
+        """Destination-only blitting operation."""
+        LOG.debug(state)
```

Comment:
can we afford these calls? I know that logging in tight loops should be avoided because the interpreter can't compile out the `LOG.debug()` call from being there and then later realizes it is not useful.

Last time I had to do something like this we would do something like `if DEBUG: <logging statement>` to avoid the more expensive function call and the interpreter could compile out the statement if DEBUG was false. It was in Perl so this would need to be reverified for python.

That said, we shouldn't do this until profiling prooves us that this is expensive.

Reply:
The `str()` representation is not evaluated if the verbosity level is not active, so these are just NOOP function calls. I did not have any trouble rendering captures with `INFO` level verbosity, and `DEBUG` was mostly bottlenecked by stderr flushing, so I don't think this is an issue. It will be trivial to fix if we ever get performance issue reports